DM-37330: Add a utility function to compress tract list

arunkannawadi opened the pull request:
This PR adds a utility function to `plotUtils` and unit tests for it. It also copies the unit tests for other `plotUtils` functions from `analysis_drp`.

leeskelvin commented on tests/test_plotUtils.py:
```diff
@@ -0,0 +1,97 @@
+
+
+class FitTest(lsst.utils.tests.TestCase):
+    """Test to see if the fitting and distance calculations are working"""
```
Missing period.

leeskelvin commented on tests/test_plotUtils.py:
```diff
@@ -0,0 +1,97 @@
+        self.assertFloatsAlmostEqual(paramsOut["bPerpMax"], 22.0)
+
+    def testPerpDistance(self):
+        """Test the calculation of the perpendicular distance"""
```
Missing period.

leeskelvin commented on python/lsst/analysis/tools/actions/plot/plotUtils.py:
```diff
@@ -598,6 +598,68 @@ def addSummaryPlot(fig, loc, sumStats, label):
+
+    Parameters
+    ----------
+    numbers : `list` [`int`]
```
Rather than ``list``, should this just be `iterable` (without back ticks)? From the doc string and the unit test, it looks like you're happy accepting any iterable?

leeskelvin commented on python/lsst/analysis/tools/actions/plot/plotUtils.py:
```diff
@@ -598,6 +598,68 @@ def addSummaryPlot(fig, loc, sumStats, label):
+    >>> shorten_list(range(4))
+    "0-3"
+    """
+    if not numbers:  # empty container
```
As we discussed:
```
if len(numbers) == 0:
```
will catch numpy array inputs.

leeskelvin replied:
Alternatively, perhaps convert to a `set` right at the top of this function. This would then allow for the code block checking for possible duplicate number entires in your input iterable to be dropped.

leeskelvin replied:
Option #3: you could swap the order of `len(numbers)` and `sorted` such that sorting occurs first. This will guarantee that your numbers iterable is in a list format, which would then have a `len`. Otherwise, the input (e.g., an iterator) may not have a `len` to pass the initial `if` check.

arunkannawadi replied:
I think I am going to go with option 3. If it is an empty container, calling `sorted` on it is a trivial operation. OTOH, creating a set out of a large container might be relatively more expensive. Option 1 is still the best for our use case, but I am going to refrain from assuming that __len__ is implemented for ``numbers``.

arunkannawadi commented:
I went with option 3, for reasons I indicated in the inline comments. I also added two keyword-only arguments that can help customize this utility function. I have updated the examples and tests to illustrate that as well.
